### PR TITLE
Flood Detection on Pi

### LINKED ISSUE #6: Flood  Detection
**Is your feature request related to a problem? Please describe.**

Create a package using lgpio to detect floods in the ebay using the blue-rov flood detection.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### LINKED ISSUE #98: [FEATURE] Investigate using cron jobs instead of robot upstart
### Is your feature request related to a problem? Please describe

It's annoying to use the robot upstart package to launch packages on the Pi when it turns on.

### Describe the solution you'd like

Investigate using cron jobs instead of robot upstart to launch those packages.

### PR DESCRIPTION
**Pull request by Tazmania40**
# Pull Request

## What type of PR is this? (check all applicable)

- [X] Refactor
- [X] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

Adding the Pi-based support for the flood detector modules.

## Related Tickets & Documents

- Closes #6 
- Closes #98

### CODE REVIEW

**Comment by InvincibleRMC** [comment]
TODO make new issue for copying service files / refeactor install file

**Comment by InvincibleRMC** [comment]
@NoahMollerstuen mentioned that valve manip also using gpio might cause a race condition.

**Comment by InvincibleRMC** [comment]
Will refactor that on in the future if that is the case

**Comment by InvincibleRMC** [comment]
Someone should add python3-lgpio to rosdep then add that as a dep.

File: src/pi/flood_detection/flood_detection/flood_detector.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        data1 = lgpio.gpio_read(self.gpio_chip, DETECT_PIN)
+
+        # If any of the sensors detect water, send true to /tether/flooding
+        msg = Flooding()
+
+        if data1 == lgpio.HIGH:
+            msg.flooding = True
+            self.get_logger().error("The ebay is flooding")
+        else:
+            msg.flooding = False
```
Comment (benjaminwp18) [inline src/pi/flood_detection/flood_detection/flood_detector.py:30]:
```
        flood_reading = lgpio.gpio_read(self.gpio_chip, DETECT_PIN)

        # If any of the sensors detect water, send true to /tether/flooding
        msg = Flooding()
        msg.flooding = (flood_reading == lgpio.HIGH)
        if msg.flooding:
            self.get_logger().error("The ebay is flooding")
```

File: src/pi/flood_detection/flood_detection/gpio_reader_no_ros.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
+import lgpio
+
+# Pins used for GPIO
+DETECT_PIN = 17
+
+
+def main() -> None:
+    # GPIO Boilerplate
+    h = lgpio.gpiochip_open(0)
```
Comment (benjaminwp18) [inline src/pi/flood_detection/flood_detection/gpio_reader_no_ros.py:9]:
```suggestion
    gpio_chip = lgpio.gpiochip_open(0)
```

File: src/pi/flood_detection/flood_detection/gpio_reader_no_ros.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    lgpio.gpio_claim_input(h, DETECT_PIN)
```
Comment (benjaminwp18) [inline src/pi/flood_detection/flood_detection/gpio_reader_no_ros.py:10]:
```suggestion
    lgpio.gpio_claim_input(gpio_chip, DETECT_PIN)
```

File: src/pi/flood_detection/flood_detection/gpio_reader_no_ros.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+    data: int = lgpio.gpio_read(h, DETECT_PIN)
```
Comment (benjaminwp18) [inline src/pi/flood_detection/flood_detection/gpio_reader_no_ros.py:12]:
```suggestion
    data: int = lgpio.gpio_read(gpio_chip, DETECT_PIN)
```

File: src/pi/flood_detection/flood_detection/gpio_reader_no_ros.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+    first_run = True
+
+    while True:
+        # Read Data
+        old_data: int = data
+
+        data = lgpio.gpio_read(h, DETECT_PIN)
```
Comment (benjaminwp18) [inline src/pi/flood_detection/flood_detection/gpio_reader_no_ros.py:20]:
```suggestion
        data = lgpio.gpio_read(gpio_chip, DETECT_PIN)
```

File: src/pi/flood_detection/package.xml [hunk 6]
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0"?>
+<?xml-model href="http://download.ros.org/schema/package_format3.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
+<package format="3">
+  <name>flood_detection</name>
+  <version>0.0.3</version>
+  <description>Detects flood</description>
+  <maintainer email="hwo2@case.edu">Henry Odza</maintainer>
```
Comment (benjaminwp18) [inline src/pi/flood_detection/package.xml:7]:
```suggestion
  <maintainer email="rmc170@case.edu">Michael Carlstrom</maintainer>
```

File: src/pi/flood_detection/flood_detection/gpio_reader_no_ros.py
Comment (benjaminwp18) [inline src/pi/flood_detection/flood_detection/gpio_reader_no_ros.py:1]: Rename to `gpio_test`

File: src/pi/flood_detection/setup.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    zip_safe=True,
+    maintainer='Michael Carlstrom',
+    maintainer_email='rmc@carlstrom.com',
+    description='PI to Pixhawk Communication',
+    license='Apache License 2.0',
+    tests_require=['pytest'],
+    entry_points={
+        'console_scripts': [
+            'flood_detector = flood_detection.flood_detector:main',
+            'test = flood_detection.gpio_reader_no_ros:main'
```
Comment (benjaminwp18) [inline src/pi/flood_detection/setup.py:29]:
```suggestion
            'test = flood_detection.gpio_test:main'
```

File: src/pi/pi_main/README.md [hunk 8]
```diff
@@ -113,34 +113,26 @@ This should automatically be done by the prior command `ros2 run pi_main install
...
-```
+Installing & setting up this package creates a startup task called `pi_main`. You can manually start and stop this task.
 
-To run the `cwrubotix_pi` task in the background (happens on Pi startup):
+To run the `pi_main` task in the background (happens on Pi startup):
 
 ```bash
-sudo systemctl start cwrubotix_pi.service
+sudo systemctl start pi_main.service
 ```
```
Comment (benjaminwp18) [inline src/pi/pi_main/README.md:125]:
````suggestion

To kill the `pi_main` background task (**do this before starting the foreground task**):
```bash
sudo systemctl stop pi_main.service
```
````

**Review by benjaminwp18 — CHANGES_REQUESTED** [review]

**Review by benjaminwp18 — APPROVED** [review]

**Pull request merged**